src/compute/src/sink/continual_task.rs

```diff
@@ -0,0 +1,793 @@
+
+            // A lower bound on the upper of the output.
+            output_progress: Antichain::from_elem(Timestamp::minimum()),
+        }
+    }
+
+    /// Returns data to write to the output, if any, and the new upper to use.
+    ///
+    /// TODO(ct): Remove the Vec allocation here.
+    fn process(&mut self) -> Option<(Antichain<Timestamp>, Vec<((&K, &V), &Timestamp, &D)>)> {
+        // We can only append at times >= the output_progress, so pop off
+        // anything unnecessary.
+        while let Some(x) = self.append_times.first() {
+            if self.output_progress.less_equal(x) {
+                break;
+            }
+            self.append_times.pop_first();
+        }
+
+        // Find the smallest append_time before append_time_progress. This is
+        // the next time we might need to write data at. Note that we can only
+        // act on append_times once the progress has passed them, because they
+        // could come out of order.
+        let write_ts = match self.append_times.first() {
+            Some(x) if !self.append_times_progress.less_equal(x) => x,
+            Some(_) | None => {
+                // The CT sink's contract is that it only writes data at times
+                // we received an input diff. There are none in
+                // `[output_progress, append_times_progress)`, so we can go
+                // ahead and advance the upper of the output, if it's not
+                // already.
+                //
+                // We could instead ensure liveness by basing this off of
+                // to_append, but for any CTs reading the output (expected to be
+                // a common case) we'd end up looping each timestamp through
+                // persist one-by-one.
+                if PartialOrder::less_than(&self.output_progress, &self.append_times_progress) {
+                    return Some((self.append_times_progress.clone(), Vec::new()));
+                }
+                // Otherwise, nothing to do!
+                return None;
+            }
+        };
+
+        // We can also advance the output upper up to the write_ts if it's not
+        // there already.
+        if self.output_progress.less_than(write_ts) {
+            return Some((Antichain::from_elem(write_ts.clone()), Vec::new()));
+        }
+
+        if self.to_append_progress.less_equal(write_ts) {
+            // Don't have all the necessary data yet.
+            return None;
+        }
+
+        // Time to write some data! Produce the collection as of write_ts by
+        // advancing timestamps, consolidating, and filtering out anything at
+        // future timestamps.
+        let as_of = &[write_ts.clone()];
+        for (_, t, _) in &mut self.to_append {
+            t.advance_by(AntichainRef::new(as_of))
+        }
+        // TODO(ct): Metrics for vec len and cap.
+        consolidate_updates(&mut self.to_append);
+        // WIP resize the vec down if cap >> len?
```

Comment:
Might be possible to use `ConsolidatingVec` to that end. That would also solve another issue I think we have now, namely that `to_append` is never consolidated unless new updates arrive in the input.

Reply:
> ConsolidatingVec has already erased the timestamps, but this needs to keep them around so it can do the filter vs write_ts below.

Oh, I was thinking you could use a `ConsolidatingVec<(Row, Timestamp)>`. Would that not work?

> Oh, Correction might be almost exactly what we need though!

Also good, though note that the MV persist sink will hopefully replace that with an implementation that can spill to disk. I'm not sure if that would also be suitable for the CT sink, it might be unnecessarily complex. But we can of course always copy the current implementation.

Reply:
Oh clever, `ConsolidatingVec<(Row, Timestamp)>` should work AFAICT. I tried quickly typing it up and it passes the unit tests and also the sqllogictests, so I think we have our answer. I shared your concerns about `Correction`.

Hooking this up requires exposing bits of `ConsolidatingVec` in a way that I'm not comfortable sneaking into a PR that's already stamped, so I'll leave the switch for a followup